### PR TITLE
Separate CArray and family from `core.pyx`

### PR DESCRIPTION
**Pull request by niboshi**
Similar refactoring to #2785
Removes `include 'carray.pxi'` from `core.pyx`.

CPU performance benchmark
#### Before
```
empty               :     0.233 us   +/- 0.177 (min:    0.188) us
sum                 :    17.517 us   +/- 1.471 (min:   16.689) us
sum_huge            :    21.874 us   +/- 4.227 (min:   18.048) us
add                 :    12.300 us   +/- 1.122 (min:   11.752) us
add_out             :     9.184 us   +/- 1.086 (min:    8.735) us
add_out_huge        :    11.666 us   +/- 2.149 (min:   10.163) us
userkernel_create   :     2.798 us   +/- 0.667 (min:    2.630) us
userkenel           :     8.929 us   +/- 0.933 (min:    8.561) us
userkernel_huge     :    10.652 us   +/- 1.015 (min:    9.890) us
```

#### This PR
```
empty               :     0.213 us   +/- 0.013 (min:    0.191) us
sum                 :    17.852 us   +/- 1.594 (min:   16.859) us
sum_huge            :    21.445 us   +/- 2.672 (min:   19.947) us
add                 :    12.286 us   +/- 1.482 (min:   11.742) us
add_out             :     9.317 us   +/- 1.137 (min:    8.825) us
add_out_huge        :    11.169 us   +/- 1.502 (min:   10.355) us
userkernel_create   :     2.887 us   +/- 0.642 (min:    2.694) us
userkenel           :     9.054 us   +/- 1.217 (min:    8.680) us
userkernel_huge     :    10.510 us   +/- 1.224 (min:    9.784) us
```

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1670,7 +1679,134 @@ cpdef vector.vector[Py_ssize_t] _get_strides_for_order_K(ndarray x, dtype,
...
+    for line in source.splitlines(keepends=True):
+        if _cucomplex_include_pattern.match(line):
+            lines += '#include <cupy/cuComplex_bridge.h>  '\
+                     '// translate_cucomplex\n'
+        else:
+            lines += line
+    return ''.join(lines)
+
+
+cpdef function.Module compile_with_cache(
```
Comment (niboshi) [inline cupy/core/core.pyx]: I think this function should not be in `core.pyx` as well, but I wonder where it should be.
`cupy/cuda/compiler.py`? But I'm not sure if this function can be uncythonized.

Comment (asi1024) [reply cupy/core/core.pyx]: I recognize that one of difference between `cuda.compile_with_cache` and this function is whether `core/include` is included in `--include-path` or not. So I think this function should be placed here.

File: cupy/core/_carray.pxd [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@cython.final
+cdef class CArray(function.CPointer):
+
+    cdef:
+        _CArray val
+
+    cdef void init(
+        self, void* data_ptr, Py_ssize_t data_size,
+        const vector.vector[Py_ssize_t]& shape,
+        const vector.vector[Py_ssize_t]& strides)
```
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/core/_carray.pxd]: I think this method does not raise any Python exceptions?

[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/_carray.pxd]: Sorry, this method does not raise any exception.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**